importer/src/bin/clip_osm.rs

```diff
@@ -32,64 +32,78 @@ fn main() -> Result<()> {
+                    if relation.members().any(|(obj_type, id, _)| {
+                        (obj_type == OSMObjectType::Node && node_ids_within_boundary.contains(&id))
+                            || (obj_type == OSMObjectType::Way && way_ids.contains(&id))
+                            || (obj_type == OSMObjectType::Relation && relation_ids.contains(&id))
+                    }) {
+                        relation_ids.insert(relation.id());
+                    }
+                }
+            }
+        }
+    }
 
+    let mut writer = osmio::xml::XMLWriter::new(BufWriter::new(File::create(out_path)?));
+    // Second Pass: write the feature for each ID accumulated in the first pass
     let mut reader = osmio::pbf::PBFReader::new(BufReader::new(File::open(pbf_path)?));
     for obj in reader.objects() {
         match obj.object_type() {
             OSMObjectType::Node => {
                 let node = obj.into_node().unwrap();
-                if node.lat_lon().is_some() {
-                    nodes.insert(node.id(), node);
+                if way_node_ids.contains(&node.id()) {
+                    writer.write_obj(&RcOSMObj::Node(node))?;
                 }
             }
             OSMObjectType::Way => {
-                // Assume all nodes appear before any way.
-                let way = obj.into_way().unwrap();
-                if way.nodes().iter().any(|id| {
-                    nodes
-                        .get(id)
-                        .map(|n| boundary.contains(&to_pt(n.lat_lon().unwrap())))
-                        .unwrap_or(false)
-                }) {
-                    ways.insert(way.id(), way);
+                if way_ids.contains(&obj.id()) {
+                    let way = obj.into_way().unwrap();
+                    writer.write_obj(&RcOSMObj::Way(way))?;
                 }
             }
             OSMObjectType::Relation => {
-                let relation = obj.into_relation().unwrap();
-                if relation.members().any(|(obj_type, id, _)| {
-                    (obj_type == OSMObjectType::Node && nodes.contains_key(&id))
-                        || (obj_type == OSMObjectType::Way && ways.contains_key(&id))
-                        || (obj_type == OSMObjectType::Relation && relations.contains_key(&id))
-                }) {
-                    relations.insert(relation.id(), relation);
+                if relation_ids.contains(&obj.id()) {
+                    let relation = obj.into_relation().unwrap();
+                    writer.write_obj(&RcOSMObj::Relation(relation))?;
                 }
             }
         }
     }
 
-    // Trim out all unused nodes
-    let mut used_nodes = HashSet::new();
-    for way in ways.values() {
-        used_nodes.extend(way.nodes().into_iter().cloned());
-    }
-
-    let mut writer = osmio::xml::XMLWriter::new(BufWriter::new(File::create(out_path)?));
-    // TODO Nondetermistic output because of HashMap!
```

Comment:
A side effect of these changes is that we now maintain order of the input objects.